src/utils.jl

```diff
@@ -31,13 +45,32 @@ julia> StructArrays._map_params(T -> Complex{T}, Tuple{Int32,Float64})
 (Complex{Int32}, Complex{Float64})
 ```
 """
-_map_params(f, ::Type{Tuple{}}) = ()
-function _map_params(f, ::Type{T}) where {T<:Tuple}
-    (f(tuple_type_head(T)), _map_params(f, tuple_type_tail(T))...)
-end
 _map_params(f::F, ::Type{NamedTuple{names, types}}) where {names, types, F} =
     NamedTuple{names}(_map_params(f, types))
 
+function _map_params(f::F, ::Type{T}) where {T<:Tuple, F}
+    if @generated
+        types = types_to_tuple(T)
+        ex = :()
+        for t ∈ types
+            push!(ex.args, :(f($t)))
+        end
+        ex
+    else
+        _map_params_recursive(f, T)
+    end
+end
+
+_map_params_recursive(f, ::Type{Tuple{}}) = ()
+function _map_params_recursive(f, ::Type{T}) where {T<:Tuple}
+    (f(tuple_type_head(T)), _map_params_recursive(f, tuple_type_tail(T))...)
+end
+
+types_to_tuple(::Type{Tuple{}}) = ()
+function types_to_tuple(::Type{T}) where {T <: Tuple}
+    (Base.tuple_type_head(T), types_to_tuple(Base.tuple_type_tail(T))...)
+end
```

Comment:
Could `types_to_tuple` just be replaced by `fieldtypes`? I think this recurrent trick was for inferrability, but now that it is just a fallback, I imagine it's best to keep the code simple (in general, I'd like to stop using `tuple_type_head` & co., as I think they're getting deprecated).

Reply:
Right, I didn't think of using `fieldtypes` this way, I'll make the changes. I agree with you about avoiding the use of `tuple_type_head` and `tuple_type_tail`.